File: shared/src/main/java/com/yuriy/openradio/model/media/item/MediaItemCountryStations.java
/*
 * Copyright 2017 The "Open Radio" Project. Author: Chernyshov Yuriy
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.yuriy.openradio.model.media.item;

import androidx.annotation.NonNull;

import com.yuriy.openradio.model.net.UrlBuilder;
import com.yuriy.openradio.utils.AppLogger;
import com.yuriy.openradio.utils.ConcurrentUtils;
import com.yuriy.openradio.vo.RadioStation;

import java.util.ArrayList;
import java.util.List;

/**
 * Created by Yuriy Chernyshov
 * At Android Studio
 * On 8/31/15
 * E-Mail: chernyshov.yuriy@gmail.com
 * <p>
 * {@link MediaItemCountryStations} is concrete implementation of the {@link MediaItemCommand} that
 * designed to prepare data to display radio stations of the single Country.
 */
public final class MediaItemCountryStations extends IndexableMediaItemCommand {

    private static final String LOG_TAG = MediaItemCountryStations.class.getSimpleName();

    /**
     * Default constructor.
     */
    public MediaItemCountryStations() {
        super();
    }

    @Override
    public void execute(final IUpdatePlaybackState playbackStateListener,
                        @NonNull final MediaItemShareObject shareObject) {
        super.execute(playbackStateListener, shareObject);
        AppLogger.d(LOG_TAG + " invoked");
        // Use result.detach to allow calling result.sendResult from another thread:
        shareObject.getResult().detach();

        ConcurrentUtils.API_CALL_EXECUTOR.submit(
                () -> {
                    // Load all categories into menu
                    final List<RadioStation> list = new ArrayList<>(
                            shareObject.getServiceProvider().getStations(
                                    shareObject.getDownloader(),
                                    UrlBuilder.getStationsInCountry(
                                            shareObject.getCountryCode(),
                                            getPageNumber() * (UrlBuilder.ITEMS_PER_PAGE + 1),
                                            UrlBuilder.ITEMS_PER_PAGE
                                    )
                            )
                    );
                    handleDataLoaded(playbackStateListener, shareObject, list);
                }
        );
    }
}

File: common/src/main/java/com/yuriy/openradio/shared/model/media/item/MediaItemCountryStations.kt
/*
 * Copyright 2016-2022 The "Open Radio" Project. Author: Chernyshov Yuriy
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *      http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.yuriy.openradio.shared.model.media.item

import com.yuriy.openradio.shared.model.media.item.MediaItemCommand.IUpdatePlaybackState
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.launch
import kotlinx.coroutines.withTimeoutOrNull

/**
 * Created by Yuriy Chernyshov
 * At Android Studio
 * On 8/31/15
 * E-Mail: chernyshov.yuriy@gmail.com
 *
 * [MediaItemCountryStations] is concrete implementation of the [MediaItemCommand] that
 * designed to prepare data to display radio stations of the single Country.
 */
class MediaItemCountryStations : IndexableMediaItemCommand() {

    override fun execute(playbackStateListener: IUpdatePlaybackState, dependencies: MediaItemCommandDependencies) {
        super.execute(playbackStateListener, dependencies)
        // Use result.detach to allow calling result.sendResult from another thread:
        dependencies.result.detach()
        if (dependencies.isSavedInstance) {
            deliverResult(dependencies)
            return
        }
        mJob?.cancel()
        mJob = dependencies.mScope.launch(Dispatchers.IO) {
            withTimeoutOrNull(MediaItemCommand.CMD_TIMEOUT_MS) {
                // Load all categories into menu
                val pageNumber = nextPageNumber
                handleDataLoaded(
                    playbackStateListener,
                    dependencies,
                    dependencies.presenter.getStationsByCountry(dependencies.countryCode, pageNumber),
                    pageNumber
                )
            } ?: dependencies.result.sendResult(null)
        }
    }
}
